### PR TITLE
[Merged by Bors] - feat: `ℂ` and `Finset.expect`

### PR DESCRIPTION
**Pull request by YaelDillies**
Make more arguments to `NNReal.coe_sum` explicit to match `NNReal.coe_expect`.

From LeanAPAP

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 1]
```diff
@@ -39,7 +40,7 @@ their counterparts in `Mathlib/Analysis/Complex/Basic.lean` (which causes linter
 A few lemmas requiring heavier imports are in `Mathlib/Data/RCLike/Lemmas.lean`.
 -/
 
-open scoped ComplexConjugate
+open scoped BigOperators ComplexConjugate
```
Comment (b-mehta) [inline Mathlib/Analysis/RCLike/Basic.lean:43]:
```suggestion
open scoped ComplexConjugate
```
somebody told me that's a no-op :)

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 2]
```diff
@@ -605,13 +610,23 @@ variable (K) in
 lemma nnnorm_nsmul [NormedAddCommGroup E] [NormedSpace K E] (n : ℕ) (x : E) :
     ‖n • x‖₊ = n • ‖x‖₊ := by simpa [Nat.cast_smul_eq_nsmul] using nnnorm_smul (n : K) x
 
+section NormedField
+variable [NormedField E] [CharZero E] [NormedSpace K E]
```
Comment (b-mehta) [inline Mathlib/Analysis/RCLike/Basic.lean:614]: do you want q and x here too?

File: Mathlib/Data/NNReal/Basic.lean [hunk 3]
```diff
@@ -282,6 +284,10 @@ theorem coe_multiset_prod (s : Multiset ℝ≥0) : ((s.prod : ℝ≥0) : ℝ) =
 theorem coe_sum {α} {s : Finset α} {f : α → ℝ≥0} : ↑(∑ a ∈ s, f a) = ∑ a ∈ s, (f a : ℝ) :=
   map_sum toRealHom _ _
 
+@[simp, norm_cast]
+lemma coe_expect {ι : Type*} (s : Finset ι) (f : ι → ℝ≥0) : 𝔼 i ∈ s, f i = 𝔼 i ∈ s, (f i : ℝ) :=
```
Comment (b-mehta) [inline Mathlib/Data/NNReal/Basic.lean:288]: Let's make the index type consistent with the above

File: Mathlib/Analysis/RCLike/Basic.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Analysis/RCLike/Basic.lean:43]: Not anymore! `expect` is scoped to it now

File: Mathlib/Analysis/RCLike/Basic.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Analysis/RCLike/Basic.lean:614]: I don't like putting explicit variables in `variable`. That always catches me off guard, except when it's clearly used throughout the section, which would not be the case here

Comment (b-mehta) [reply Mathlib/Analysis/RCLike/Basic.lean:614]: Interesting personal preference, but okay

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 4]
```diff
@@ -605,13 +610,23 @@ variable (K) in
...
+  simpa [NNRat.cast_smul_eq_nnqsmul] using norm_smul (q : K) x
 
 variable (K) in
-lemma nnnorm_nnqsmul [NormedField E] [CharZero E] [NormedSpace K E] (q : ℚ≥0) (x : E) :
-    ‖q • x‖₊ = q • ‖x‖₊ := by simpa [NNRat.cast_smul_eq_nnqsmul] using nnnorm_smul (q : K) x
+lemma nnnorm_nnqsmul (q : ℚ≥0) (x : E) : ‖q • x‖₊ = q • ‖x‖₊ := by
+  simpa [NNRat.cast_smul_eq_nnqsmul] using nnnorm_smul (q : K) x
+
+@[bound]
+lemma norm_expect_le {ι : Type*} {s : Finset ι} {f : ι → E} : ‖𝔼 i ∈ s, f i‖ ≤ 𝔼 i ∈ s, ‖f i‖ :=
```
Comment (b-mehta) [inline Mathlib/Analysis/RCLike/Basic.lean:626]: Is RCLike really the right generality for norm_expect_le?

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 5]
```diff
@@ -605,13 +610,23 @@ variable (K) in
...
+  Finset.le_expect_of_subadditive norm_zero norm_add_le (fun _ _ ↦ by rw [RCLike.norm_nnqsmul K])
```
Comment (b-mehta) [inline Mathlib/Analysis/RCLike/Basic.lean:627]:
```suggestion
  Finset.le_expect_of_subadditive norm_zero norm_add_le fun _ _ ↦ by rw [norm_nnqsmul K]
```

File: Mathlib/Data/NNReal/Basic.lean (hunk 3 above)
Comment (b-mehta) [reply Mathlib/Data/NNReal/Basic.lean:288]: (and below!)

File: Mathlib/Analysis/RCLike/Basic.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Analysis/RCLike/Basic.lean:626]: Perhaps surprisingly: yes

A real or complex vector space is the most natural (and almost only) kind of spaces where `‖n • x|| = n • ‖x‖` for `n` a natural number (which in turn gives `‖q • x|| = |q| • ‖x‖` for `q` rational). In general, we only have `‖n • x|| ≤ n • ‖x‖`.

File: Mathlib/Data/NNReal/Basic.lean (hunk 3 above)
Comment (b-mehta) [reply Mathlib/Data/NNReal/Basic.lean]: below once more!

File: Mathlib/Data/Complex/BigOperators.lean [hunk 6]
```diff
@@ -23,12 +24,24 @@ theorem ofReal_prod (f : α → ℝ) : ((∏ i ∈ s, f i : ℝ) : ℂ) = ∏ i
 theorem ofReal_sum (f : α → ℝ) : ((∑ i ∈ s, f i : ℝ) : ℂ) = ∑ i ∈ s, (f i : ℂ) :=
   map_sum ofReal _ _
 
+@[simp, norm_cast]
+lemma ofReal_expect (f : α → ℝ) : 𝔼 i ∈ s, f i = 𝔼 i ∈ s, (f i : ℂ) :=
```
Comment (b-mehta) [inline Mathlib/Data/Complex/BigOperators.lean:28]: Let's add type ascriptions here, to match that previous two, and for readability.

**Review by b-mehta — APPROVED** [review]
Nice!

bors merge

**Pull request closed without merging**